### DISCUSSION TITLE
Where and how are the constraints on the energy minimization imposed?

### DISCUSSION THREAD

**Discussion by sitangshugk95** [ContactEngineering/ContactMechanics#103] [Q&A] [answered]
I understand that system.mininize_proxy() minimizes the energy of the system, using either CG/L-BGFS-B 
etc. I am not being able to find which function imposes the non-penetration (g>=0) and average pressure/displacements constraints (for the primal/dual problem). 

Could someone please point me to the respective code / function that does this?

**Comment by pastewka** [comment]
Hi @sitangshugk95, we have an implementation of the constrained CG algorithm of Polonsky & Keer that you can find here:

https://github.com/ContactEngineering/ContactMechanics/blob/master/ContactMechanics/Optimization/ConstrainedConjugateGradients.py

That algorithm **does not** work with adhesion and it the default for `NonSmoothContactSystem` (part of the `ContactMechanics` project):

https://github.com/ContactEngineering/ContactMechanics/blob/b152466785431a8856795ec99561ef1bf4a1c94f/ContactMechanics/Systems.py#L532

For adhesive contact problems with a constrained, I recommend using L-BFGS-B -- which is implemented in `scipy` and not `ContactMechanics` -- and is the default for the `SmoothContactSystem` (part of the `Adhesion` project). You need to explicitly pass `lbounds` to this minimizer if you want a constrained minimization. The `lbounds` parameters is passed on to `scipy`.

I don't know how to switch between primal and dual problem, @sannant implemented this and may want to weight in here.

Note that the whole optimization is quite convoluted. @sannant and myself discussed multiple times that the code should be reorganized to make it easier to use.

**Reply by sitangshugk95** [reply]
Thanks for clearing my confusion, @pastewka . Since most of the papers in literature talk about solving the contact mechanics between rough surfaces as essentially minimizing the strain energy of the system subject to non-penetration/total load constraints, I was trying to go through the code and see how the constraints were implemented. I have a couple of related follow-up questions:

1. Why is L-BGFS-B not used for Contact Mechanics by default (and CCG is preferred instead)?
2. Why does CCG not work for Adhesion?

**Reply by pastewka** [reply]
1. You should be able to use L-BGFS-B for nonadhesive contact. We have never tested whether this is actually faster than CCG, but I don't think so.
2. You cannot use CCG as is for adhesion because it is intrinsically solving a linear system (the CG part), while adhesion adds additional nonlinearity (on top of the constraint) that the optimizer needs to handle.

**Reply by sannant** [reply, marked as answer]
For completeness: 

CCG can be used to some extent for the adhesive problem:
https://link.springer.com/article/10.1007/s11249-017-0980-z

In fact, we have used it in https://link.springer.com/article/10.1007/s11249-021-01454-6 , where the adhesion is close to the DMT limit. 

But it breaks down when the adhesion is strong and short ranged. 

L-BDFGS-B is definitely the goto method. We only used CCG because we had a parallel implementation of CCG but not L-BFGS-B

**Comment by sannant** [comment]
Here is maybe a more transparent way to use the scipy optimizer 

https://github.com/ContactEngineering/Adhesion/blob/master/docs/examples/hardwall_adhesion_custom.py

where a few changes ("primal -> dual" and contact forces for initial guess) could allow to use the dual approach. If you really want to do that I can also give you an example.

**Reply by pastewka** [reply]
I agree that this is more transparent than using the proxy.

**Reply by sitangshugk95** [reply]
Thanks for the link @sannant 
I think this should suffice for now. I'll go through this and let you know if I need an example.
